lib/charms/rolling_ops/v0/rollingops.py

```diff
@@ -0,0 +1,353 @@
+        """
+        params:
+            charm: the charm we are attaching this to.
+            rel_name: an identifier, by convention based on the name of the relation in the
+                metadata.yaml, which identifies this instance of RollingOperatorsFactory,
+                distinct from other instances that may be hanlding other events.
+            func_run: a closure to run when we have a lock. (It must take a CharmBase object and
+                EventBase object as args.)
+            event_acquire: The event that will prompt us to acquire a lock.
+        """
+        super().__init__(charm, None)
+
+        self.charm = charm
+        self.name = rel_name
+        self._func_run = func_run
+
+        self.framework.observe(event_acquire, self._on_acquire_lock)
+        self.framework.observe(charm.on[self.name].relation_changed, self._on_relation_changed)
+        self.framework.observe(charm.on.run_with_lock, self._on_run_with_lock)
+        self.framework.observe(charm.on.process_locks, self._on_process_locks)
+
+    def _func_run(self: CharmBase, event: EventBase) -> None:
+        """Placeholder for the function that actually runs our event.
+
+        Usually overridden in the init.
+        """
+        raise NotImplementedError
+
+    def _on_relation_changed(self: CharmBase, event: RelationChangedEvent):
+        """Process relation changed.
+
+        First, determine whether this unit has been granted a lock. If so, emit a RunWithLock
+        event.
+
+        Then, if we are the leader, fire off a process locks event.
+
+        """
+        if not event.relation.name == self.name:
+            # This is not relevant to us.
+            return
+
+        if Lock(self.name, self.charm).is_held():
+            self.charm.on.run_with_lock.emit(name=self.name)
+
+        if self.framework.model.unit.is_leader():
+            self.charm.on.process_locks.emit(name=self.name)
+
+    def _on_process_locks(self: CharmBase, event: ProcessLocks):
+        """Process locks.
+
+        Runs only on the leader. Updates the status of all locks.
+
+        """
+        if not event.name == self.name:
+            return
+
+        if not self.framework.model.unit.is_leader():
+            return
+
+        pending = []
+
+        for lock in Locks(self.name, self.charm):
+            if lock.is_held():
+                # One of our units has the lock -- return without further processing.
+                return
+
+            if lock.release_requested():
+                lock.clear()  # Updates relation data
+
+            if lock.is_pending():
+                if lock.unit == self.charm.model.unit:
+                    # Always run on the leader last.
+                    pending.insert(0, lock)
+                else:
+                    pending.append(lock)
+
+        # If we reach this point, and we have pending units, we want to grant a lock to
+        # one of them.
+        if pending:
```

Comment:
Would it be useful having some kind of "rolling process finished" func that the charm also passes in that we call after all the locks have been processed (e.g. called in an else condition for this if)?

Reply:
As written, we don't have a way of distinguishing old and new locks. A lock starts in "idle" and ends in "idle". The nice thing about that is that we don't need any logic to handle the case where an operator gets impatient and sends multiple requests to kick off the restart.

The not nice thing is that it's harder to send an "all done!" message, and also harder to trigger the start of a rolling restart without running an action.

I think that we might be heading toward timestamps. I just wanted to start simple, just in case we could get away with simple :-)

Reply:
Heh. I just suggested the same thing above -- I think the juju-y way to handle this would be to emit an event signalling that all units have held the lock and run the closure -- then authors could decide (or not) to subscribe to it and respond appropriately in their charms.